File: 35.search-insert-position.go
/*
 * @lc app=leetcode id=35 lang=golang
 *
 * [35] Search Insert Position
 Given a sorted array of distinct integers and a target value, return the index if the target is found. If not, return the index where it would be if it were inserted in order.

You must write an algorithm with O(log n) runtime complexity.
Example 1:
Input: nums = [1,3,5,6], target = 5
Output: 2
*/

// @lc code=start
func searchInsert(nums []int, target int) int {

	//用二分法
	left, right := 0, len(nums)-1

	for left <= right {
		mid := left + (right-left)/2

		if nums[mid] == target {
			return mid
			//目標比較大,往右邊找
		} else if nums[mid] < target {
			left = mid + 1
		} else {
			right = mid - 1
		}
	}

	return left

	// left, right := 0, len(nums)-1

	// for left <= right {
	//     mid := left + (right-left)/2 // 防止整數溢出

	//     if nums[mid] == target {
	// 	return mid // 找到目標，返回索引
	//     } else if nums[mid] < target {
	// 	left = mid + 1 // 縮小範圍到右半部分
	//     } else {
	// 	right = mid - 1 // 縮小範圍到左半部分
	//     }
	// }

	// return left // 當 left 超過 right 時，left 是插入位置
}

// @lc code=end

File: 35.search-insert-position.py
#
# @lc app=leetcode id=35 lang=python
#
# [35] Search Insert Position
#

# @lc code=start
class Solution(object):
    def searchInsert(self, nums, target):
        """
        :type nums: List[int]
        :type target: int
        :rtype: int
        """

        # Using binary search
        # Time complexity: O(log n)
        #Initialize left pointer and right pointer

        left = 0
        right = len(nums) - 1

        # 當右指針沒有小於左指針時,迴圈執行
        # Execute the loop while the right pointer is not less than the left pointer
        while left <= right:
            # 從中間開始往兩邊找,左右指針的中間值除以2就是中間值
            # Find the middle using (left + right) / 2 and expand outward
            mid = (left + right) // 2
            # 如果中間值是target,直接返回中間值
            # Return mid if it matches the target
            if nums[mid] == target:
                return mid
            # 如果中間值小於target,表示在target在mid的右邊範圍,所以left pointer 可以直接移到 mid + 1的位置
            # If mid < target, search in the right half by moving left to mid  + 1
            elif nums[mid] < target:
                left = mid + 1
            # 如果中間值大於target,表示在target在mid的左邊範圍,所以right pointer 可以直接移到 mid - 1的位置
            # If mid  > target, search in the right half by moving right to mid - 1
            elif nums[mid] > target:
                right = mid -1 
        # 因為是以左指針的為基準,如果最後不存在左指針的超過右指針停下來,停下來的位置就是應該插入的位置
        # If the target is not found, left surpass right and indicates the insertion index
        return left

        
# @lc code=end

